Data/ByteString/Internal.hs

```diff
@@ -783,8 +796,72 @@ isSpaceChar8 :: Char -> Bool
 isSpaceChar8 = isSpaceWord8 . c2w
 {-# INLINE isSpaceChar8 #-}
 
+------------------------------------------------------------------------
+
+-- | The type of exception raised by 'overflowError'
+-- and on failure by overflow-checked arithmetic operations.
+newtype SizeOverflowException
+  = SizeOverflowException String
+
+instance Show SizeOverflowException where
+  show (SizeOverflowException err) = err
+
+instance Exception SizeOverflowException
+
+-- | Raises a 'SizeOverflowException',
+-- with a message using the given function name.
 overflowError :: String -> a
-overflowError fun = error $ "Data.ByteString." ++ fun ++ ": size overflow"
+overflowError fun = throw $ SizeOverflowException msg
+  where msg = "Data.ByteString." ++ fun ++ ": size overflow"
+
+-- | Add two non-negative numbers.
+-- Calls 'overflowError' on overflow.
+checkedAdd :: String -> Int -> Int -> Int
+checkedAdd fun x y
+  | r >= 0    = r
+  | otherwise = overflowError fun
+  where r = assert (min x y >= 0) $ x + y
+{-# INLINE checkedAdd #-}
+
+-- | Multiplies two non-negative numbers.
+-- Calls 'overflowError' on overflow.
+checkedMul :: String -> Int -> Int -> Int
+checkedMul fun !x@(I# x#) !y@(I# y#) = assert (min x y >= 0) $
+#if TIMES_INT_2_AVAILABLE
+  case timesInt2# x# y# of
+    (# 0#, _, result #) -> I# result
+    _ -> overflowError fun
+#else
+  case timesWord2# (int2Word# x#) (int2Word# y#) of
+    (# hi, lo #) -> case word2Int# (or# hi (uncheckedShiftRL# lo shiftAmt)) of
+      0# -> I# (word2Int# lo)
+      _  -> overflowError fun
+  where !(I# shiftAmt) = finiteBitSize (0 :: Word) - 1
+#endif
+
+checkedToInt :: Integral t => t -> Maybe Int
+{-# RULES
```

Comment:
I'm sorry, that's way too much. Please keep it simple, it's not worth to maintain such amount of code for a tiny feature. If `timesWord2#` gets verbose to wire in, just stick to division, as I suggested initially.

Reply:
This mess is to check for overflow in the conversion to `Int` without using an intermediate `Integer` for common types. I'm not sure how much maintenance burden it really imposes, but it's certainly very unsightly.  The rules should remain identical, except for their types. I can't unify them because any polymorphic conversion will have an extra type `Bounded` or `Bits` type constraint.

The options as I see them:

1. Remove this ugliness and just use `fromIntegral`. Simple, clean, fast. But there will be arguably wrong outputs for overflowing conversions, with no way to avoid this.
2. Keep this ugliness. For now it still has arguably wrong output for `stimes hugePositiveInteger empty` but this is fixable by changing the `Nothing` branch in `stimesPolymorphic`.
3. Go back to using intermediate an `Integer` for the conversion. Simple, clean. It's a bit slower for many types, and may necessitate a heap allocation. But do we really care?

Reply:
I'd like to avoid arithmetic operations over `Integer`. I'd do the following:

1. Because of `Integral` constraint, there is no other way around except casting `nRaw` to `Integer` first.
2. Then use this form of cast to `Int` (it compiles to an efficient Core AFAIR):
https://github.com/haskell/bytestring/blob/c7cba3977f99547b3774f8d86d28f3d591ad34f7/Data/ByteString/Builder/ASCII.hs#L275
3. Finally check that ``len < maxBound `quot` nRaw``.

Reply:
I agree that we should simplify. @Bodigrim's suggestion looks good to me. I doubt that the heap allocations for the intermediate `Integer`'s will hurt much in the context of `stimes` which will nearly always perform other heap allocations.

Reply:
@sjakobi I hadn't seen this comment earlier, sorry about that.

> @Bodigrim's suggestion looks good to me.

I'm a bit lost here. I already incorporated that suggestion months ago; previously the fallback via-Integer code-path did comparisons against `toInteger (maxBound :: Int)` and  `toInteger (maxBound :: Int)`, and was a fair amount slower as a result. (I reduced the `RULES` to only cover `Int` and `Word` instead of every `Integral` instance in base as well.)


> I doubt that the heap allocations for the intermediate `Integer`'s will hurt much in the context of `stimes` which will nearly always perform other heap allocations.

I added some benchmark code to test this claim. My machine isn't great for fine-tuning benchmarks of this sort, but removing the intermediate-Integer-avoidance resulted in a slowdown of around 15-20% for small inputs to `stimes` given an `Int` argument. That's not _terrible_, but I think users who explicitly use `Int` or `Word` deserve better.

That said, it looks like due to [upstream developments](https://gitlab.haskell.org/ghc/ghc/-/issues/20361), these rewrite rules may not be necessary to avoid intermediate `Integer`s in future releases of GHC. I wasn't yet able to build a recent development version of GHC to investigate further, though.